src/macaron/repo_finder/provenance_extractor.py

```diff
@@ -355,3 +358,354 @@ def check_if_repository_purl_and_url_match(url: str, repo_purl: PackageURL) -> b
+    expected_build_type = "https://witness.testifysec.com/attestation-collection/v0.1"
+
+    #: Determines the expected ``attestations.type`` field in the Witness provenance predicate.
+    expected_attestation_type = "https://witness.dev/attestations/gitlab/v0.1"
+
+    def get_build_invocation(self, statement: InTotoV01Statement | InTotoV1Statement) -> tuple[str | None, str | None]:
+        """Retrieve the build invocation information from the given statement.
+
+        This method is intended to be implemented by subclasses to extract
+        specific invocation details from a provenance statement.
+
+        Parameters
+        ----------
+        statement : InTotoV1Statement | InTotoV01Statement
+            The provenance statement from which to extract the build invocation
+            details. This statement contains the metadata about the build process
+            and its associated artifacts.
+
+        Returns
+        -------
+        tuple[str | None, str | None]
+            A tuple containing two elements:
+            - The first element is the build invocation entry point (e.g., workflow name), or None if not found.
+            - The second element is the invocation URL or identifier (e.g., job URL), or None if not found.
+        """
+        if statement["predicate"] is None:
+            return None, None
+
+        attestation_type = json_extract(statement["predicate"], ["attestations", "type"], str)
+        if self.expected_attestation_type != attestation_type:
+            return None, None
+        gl_workflow = json_extract(statement["predicate"], ["attestations", "attestation", "ciconfigpath"], str)
+        gl_job_url = json_extract(statement["predicate"], ["attestations", "attestation", "joburl"], str)
+        return gl_workflow, gl_job_url
+
+
+class ProvenancePredicate:
+    """Class providing utility methods for handling provenance predicates.
+
+    This class contains static methods for extracting information from predicates in
+    provenance statements related to various build definitions. It serves as a helper
+    for identifying build types and finding the appropriate build definitions based on the extracted data.
+    """
+
+    @staticmethod
+    def get_build_type(statement: InTotoV1Statement | InTotoV01Statement) -> str | None:
+        """Extract the build type from the provided provenance statement.
+
+        Parameters
+        ----------
+        statement : InTotoV1Statement | InTotoV01Statement
+            The provenance statement from which to extract the build type.
+
+        Returns
+        -------
+        str | None
+            The build type if found; otherwise, None.
+        """
+        if statement["predicate"] is None:
+            return None
+
+        if build_type := json_extract(statement["predicate"], ["buildType"], str):
+            return build_type
+
+        return json_extract(statement["predicate"], ["buildDefinition", "buildType"], str)
```

Comment:
It might be good to document why we have two different way to extract the build type
- `json_extract(statement["predicate"], ["buildType"], str)`
- `json_extract(statement["predicate"], ["buildDefinition", "buildType"], str)`

Reply:
See commit [17f3453](https://github.com/oracle/macaron/pull/872/commits/17f3453c4be75784f00a422b71ea829e3f04e4de).

Reply:
Just for confirmation: `json_extract(statement["predicate"], ["buildDefinition", "buildType"], str)` is for `InTotoV1Statement` and `json_extract(statement["predicate"], ["buildType"], str)` is for `InTotoV01Statement` right?

Reply:
Yes in this case, but a provenance, e.g., witness with a particular version of statement, e.g., `InTotoV01Statement` might contain `buildType` in a different access path. So, the statement version does not necessarily determine the location of `buildType` .